guides-markdown/cltv.md

```diff
@@ -0,0 +1,616 @@
+const receiveValue = coin.value - 1000;
+mtx.addOutput(receiveAddr, receiveValue);
+
+// So now we have an mtx with the right input and output
+// but our input still hasn't been signed
+console.log('mtx:', mtx);
+```
+
+### Script and Sign the Inputs
+We're going to create two custom methods to handle this part
+since we can use it for the live transactions as well. The `MTX`
+primitive in bcoin has its own version of these (see the code
+[here](https://github.com/bcoin-org/bcoin/blob/master/lib/primitives/mtx.js))
+but out of the box they can only handle standard transaction types.
+
+#### scriptInput
+Roughly speaking what `scriptInput` does is take the index of the input
+you're scripting, the coin you're creating the input from, and
+the keyring used to redeem it. It then templates the input (or witness
+in this case), adding the full redeem script needed to verify the script hash
+and putting OP_0 in the place of the signature.
+
+The actual method has more checks in it to see what kind of input,
+(P2PKH, P2PK, P2SH, etc.) it is scripting for.
+
+```javascript
+/* script the inputs w/ our custom script for an mtx
+ * This is modeled after the scriptInput method on
+ * the `MTX` class
+ * @param {MTX} mtx with unscripted input
+ * @param {Number} index - index of input to script
+ * @param {Coin} coin- UTXO we are spending
+ * @param {KeyRing} ring - keyring we are signing with
+ * @returns {MTX}
+*/
+function scriptInput(mtx, index, coin, ring) {
+  const input = mtx.inputs[index];
+  const prev = coin.script;
+  const wsh = prev.getWitnessScripthash();
+  assert(ring instanceof KeyRing, 'Must pass a KeyRing to scriptInput');
+  // this is the redeem script used to verify the p2wsh hash
+  wredeem = ring.getRedeem(wsh);
+
+  assert(wredeem, 'keyring has no redeem script');
+
+  const vector = new Stack();
+
+  // first add empty space in stack for signature and public key
+  vector.pushInt(0);
+  vector.pushInt(0);
+
+  // add the raw redeem script to the stack
+  vector.push(wredeem.toRaw());
+
+  input.witness.fromStack(vector);
+  mtx.inputs[index] = input;
+  return mtx;
+}
+```
+
+#### signInput
+`signInput` takes the same parameters, simply verifies the input/witness,
+replaces the OP_0 with a signature to satisfy the script.
+
+This is a simplified version of the version of `signInput` that is on
+the `MTX` class in bcoin.
+
+```javascript
+/* This is modeled after the signInput method on
+ * the `MTX` class
+ * @param {MTX} mtx with unscripted input
+ * @param {Number} index - index of input to script
+ * @param {Coin} coin- UTXO we are spending
+ * @param {KeyRing} ring - keyring we are signing with
+ * @returns {MTX}
+*/
+function signInput(mtx, index, coin, ring) {
+  const input = mtx.inputs[index];
+  let witness, version;
+
+  const redeem = input.witness.getRedeem();
+
+  assert(
+    redeem,
+    'Witness has not been templated'
+  );
+
+  witness = input.witness;
+  version = 1;
```

Comment:
maybe indicate that this version number is for sighash type (segwit vs legacy -- and not "witness program version") I had to look it up :-)